File: algorithmic_toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.go
package main

import (
	"C"
	"fmt"
	"sort"
)

type Item struct {
	weight     int
	value      int
	unit_value float64
}

func compare(a, b Item) bool {
	return a.unit_value > b.unit_value
}

func get_optimal_value(capacity int, weights, values []int) float64 {
	weight := 0
	value := 0.0

	items := make([]Item, len(weights))
	for i := range weights {
		items[i] = Item{weights[i], values[i], float64(values[i]) / float64(weights[i])}
	}

	sort.Slice(items, func(i, j int) bool {
		return compare(items[i], items[j])
	})

	for _, item := range items {
		if weight >= capacity {
			break
		}

		units := item.weight
		if item.weight > capacity-weight {
			units = capacity - weight
		}
		weight += units
		value += float64(units) * item.unit_value
	}

	return value
}

func main() {
	var n, capacity int
	fmt.Scan(&n, &capacity)
	values := make([]int, n)
	weights := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&values[i], &weights[i])
	}

	optimal_value := get_optimal_value(capacity, weights, values)

	fmt.Printf("%.10f\n", optimal_value)
}

File: algorithmic_toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.cpp
#include <algorithm>
#include <iostream>
#include <vector>

struct Item {
  int weight;
  int value;
  double unit_value;
};

bool compare(const Item& a, const Item& b) {
  return a.unit_value > b.unit_value;
}

double get_optimal_value(int capacity, const std::vector<int> weights, const std::vector<int> values) {
  int weight = 0;
  double value = 0.0;

  std::vector<Item> items;
  for (size_t i = 0; i < weights.size(); i++)
    items.push_back({weights[i], values[i], (double)values[i] / weights[i]});

  std::sort(items.begin(), items.end(), compare);

  for (auto& item : items) {
    if (weight >= capacity) break;

    int units = std::min(item.weight, capacity - weight);
    weight += units;
    value += units * item.unit_value;
  }

  return value;
}

int main() {
  int n;
  int capacity;
  std::cin >> n >> capacity;
  std::vector<int> values(n);
  std::vector<int> weights(n);
  for (int i = 0; i < n; i++) {
    std::cin >> values[i] >> weights[i];
  }

  double optimal_value = get_optimal_value(capacity, weights, values);

  std::cout.precision(10);
  std::cout << optimal_value << std::endl;
  return 0;
}
